File: java/IBAutomater/src/ibautomater/TwoFactorAuthenticationMethod.java
/*
 * QUANTCONNECT.COM - Democratizing Finance, Empowering Individuals.
 * IBAutomater v1.0. Copyright 2019 QuantConnect Corporation.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 */

package ibautomater;

/**
 * The supported IB two-factor authentication methods.
 */
enum TwoFactorAuthenticationMethod {
    IB_KEY("ib-key"),
    MOBILE_AUTHENTICATOR("mobile-authenticator");

    private final String token;

    TwoFactorAuthenticationMethod(String token) {
        this.token = token;
    }

    /**
     * Parses a configuration token.
     *
     * @param value The token to parse
     * @return The corresponding authentication method
     */
    static TwoFactorAuthenticationMethod parse(String value) {
        String candidate = value == null ? "" : value.trim();
        if (candidate.length() == 0 || IB_KEY.token.equalsIgnoreCase(candidate)) {
            return IB_KEY;
        }
        if (MOBILE_AUTHENTICATOR.token.equalsIgnoreCase(candidate)) {
            return MOBILE_AUTHENTICATOR;
        }
        throw new IllegalArgumentException("Unsupported two-factor authentication method");
    }

    /**
     * Gets the canonical configuration token.
     *
     * @return The canonical configuration token
     */
    String getToken() {
        return this.token;
    }
}

File: QuantConnect.IBAutomater/TwoFactorAuthenticationMethod.cs
/*
 * QUANTCONNECT.COM - Democratizing Finance, Empowering Individuals.
 * IBAutomater v1.0. Copyright 2019 QuantConnect Corporation.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

namespace QuantConnect.IBAutomater
{
    /// <summary>
    /// The Interactive Brokers two-factor authentication method
    /// </summary>
    public enum TwoFactorAuthenticationMethod
    {
        /// <summary>
        /// IB Key seamless authentication through the IBKR mobile application
        /// </summary>
        IbKey = 0,

        /// <summary>
        /// A time-based one-time password generated from an authenticator setup key
        /// </summary>
        MobileAuthenticator = 1
    }
}
